### PR TITLE
Checkpoint mapped reads

### PR DESCRIPTION
**Pull request by joverlee521**
The checkpoint has been updated to also check if the sample has mapped to all segments of the reference genome. This is necessary because `samtools mpileup` does not report coverage information for segments that have no reads mapped to them, and we were getting consensus genomes that have segments that were copies of the reference genome.

### CODE REVIEW

File: scripts/checkpoint_mapped_reads.py [hunk 1]
```diff
@@ -22,11 +22,16 @@ def determine_if_all_segments_aligned(bamstats, reference) -> bool:
...
+        # try to skip header line
+        try:
+            next(f)
+        except StopIteration as err:
+            print("No coverage statistics found in BAMStats log.")
+        else:
+            for line in f:
+                bamstats_segments.add(line.split()[0])
+        finally:
+            f.close()
```
Comment (kairstenfay) [inline scripts/checkpoint_mapped_reads.py:33]: Something I just learned from @tsibley : You don't actually need to write `f.close()` when you have a `with ... as f:` statement at the top; `with` as a context manager automatically closes the declared variable `f` for you, even on error!

File: README.md [hunk 2]
```diff
@@ -157,11 +157,12 @@ Generates statistics for coverage using [BAMStats](http://bamstats.sourceforge.n
...
 
-Calculates the minimum number of reads required to meet the minimum coverage depth set for `varscan mpileup2snp` and checks the number of mapped reads from the bowtie2 alignment summary. The minimum number of reads required is calculated as:
+Determines if all segments within the reference genome has coverage summary generated from BAMStats.
+Then calculates the minimum number of reads required to meet the minimum coverage depth set for `varscan mpileup2snp` and checks the number of mapped reads from the bowtie2 alignment summary. The minimum number of reads required is calculated as:
 ```
 min_reads = (reference_genome_length * min_cov) / (raw_read_length)
 ```
-If `mapped_reads` is greater than or equal to `min_reads`, then continue with process to build consensus genome. If not, then stop the process for this sample/reference pair.
+If not all segments are represented or if `mapped_reads` is less than `min_reads`, then stop the process for this sample/reference pair.
```
Comment (kairstenfay) [inline README.md:166]: I think it's best if we squash the README changes to be in the same commit as the checkpoint changes.

File: scripts/checkpoint_mapped_reads.py (hunk 1 above)
Comment (joverlee521) [reply scripts/checkpoint_mapped_reads.py:33]: Oh! Good to know!

**Comment by joverlee521** [comment]
Once we merge this branch, I will wipe the consensus genomes and genomic sequences from the production database, and re-backfill that data. I believe that would be easier than manually deleting the "wrong" genomes on production.

File: Snakefile [hunk 3]
```diff
@@ -100,9 +100,10 @@ def aggregate_input(wildcards):
     """
     with open(checkpoints.mapped_reads.get(sample=wildcards.sample, reference=wildcards.reference).output[0]) as f:
         summary = f.readlines()
-        min_reads = float(summary[0].split()[-1])
-        mapped = float(summary[1].split()[-1])
-        if mapped <= min_reads:
+        all_segments_aligned = eval(summary[0].split()[-1])
```
Comment (tsibley) [inline Snakefile]: The use of `eval` makes me suspect about other ways of doing this. After a quick skim, is what's happening here a boolean is getting serialized to the text file as a string and you want to get it back to a boolean?

I think a better approach would be to serialize the data you want (including the floats below) to a JSON document.

**Review by tsibley — CHANGES_REQUESTED** [review]

File: Snakefile (hunk 3 above)
Comment (joverlee521) [reply Snakefile]: Yup, that's exactly what's happening. Definitely better to dump to a JSON document!

**Pull request merged**